### PR TITLE
Remove coupon code infrastructure

### PR DESCRIPTION
**Pull request by mbj**

### CODE REVIEW

**Comment by mbj** [comment]
@dkubb This is ready for review.

Scope is the removal of `Promotion#code` and everything that depends on it.

File: core/app/models/spree/promotion.rb [hunk 1]
```diff
@@ -28,22 +28,16 @@ class Promotion < Spree::Base
...
     before_save :normalize_blank_values
 
-    scope :coupons, ->{ where("#{table_name}.code IS NOT NULL") }
-
     join_table = reflect_on_association(:orders).join_table
 
     scope :applied, -> { where("EXISTS (SELECT * FROM #{join_table} WHERE #{join_table}.promotion_id = #{table_name}.id)").uniq }
 
-    self.whitelisted_ransackable_attributes = ['code', 'path', 'promotion_category_id']
+    self.whitelisted_ransackable_attributes = ['path', 'promotion_category_id']
```
Comment (dkubb) [inline core/app/models/spree/promotion.rb]: This should probably use `%w[...]`

File: api/app/controllers/spree/api/promotions_controller.rb [hunk 2]
```diff
@@ -19,7 +19,7 @@ def requires_admin
         end
 
         def load_promotion
-          @promotion = Spree::Promotion.find_by_id(params[:id]) || Spree::Promotion.with_coupon_code(params[:id])
+          @promotion = Spree::Promotion.find_by_id(params[:id])
```
Comment (dkubb) [inline api/app/controllers/spree/api/promotions_controller.rb]: Should this use `params.fetch(:id)` ?

Comment (dkubb) [inline api/app/controllers/spree/api/promotions_controller.rb]: Also, `find_by_id` is not quite as strict as normal `find`. It will accept `nil` and just do a search for `id IS NULL`. I don't know the context of this call, so maybe `params[:id]` can legitimately be `nil`, but I wanted to flag it to be sure.

File: api/spec/controllers/spree/api/promotions_controller_spec.rb [hunk 3]
```diff
@@ -22,7 +22,7 @@ module Spree
       stub_authentication!
     end
 
-    let(:promotion) { create :promotion, code: '10off' }
+    let(:promotion) { create :promotion }
```
Comment (dkubb) [inline api/spec/controllers/spree/api/promotions_controller_spec.rb]: This should probably be parenthesized.

File: backend/spec/features/admin/promotion_adjustments_spec.rb [hunk 4]
```diff
@@ -3,7 +3,7 @@
 describe "Promotion Adjustments", :type => :feature do
   stub_authorization!
 
-  context "coupon promotions", :js => true do
+  context "creating promotions", :js => true do
```
Comment (dkubb) [inline backend/spec/features/admin/promotion_adjustments_spec.rb]: This string should be single quoted.

File: backend/spec/features/admin/promotion_adjustments_spec.rb [hunk 5]
```diff
@@ -145,7 +79,7 @@
       expect(first_action_calculator.preferred_percent).to eq(10)
     end
 
-    pending "should allow an admin to create an automatic promotion with free shipping (no code)" do
+    pending "should allow an admin to create an automatic promotion with free shipping" do
```
Comment (dkubb) [inline backend/spec/features/admin/promotion_adjustments_spec.rb]: Instance of https://github.com/MountainRoseHerbs/spree/pull/210/files#r60994291

File: core/app/models/spree/promotion.rb [hunk 6]
```diff
@@ -180,7 +174,7 @@ def blacklisted?(promotable)
     end
 
     def normalize_blank_values
-      [:code, :path].each do |column|
+      [:path].each do |column|
```
Comment (dkubb) [inline core/app/models/spree/promotion.rb]: This should use `%i[...]`.

Comment (dkubb) [inline core/app/models/spree/promotion.rb]: Actually, since there is only a single attribute, I would probably just inline the code and remove the loop.

File: core/lib/spree/permitted_attributes.rb [hunk 7]
```diff
@@ -34,7 +34,7 @@ module PermittedAttributes
     ].freeze
 
     @@checkout_attributes = [
-      :coupon_code, :email, :shipping_method_id, :special_instructions, :use_billing
+      :email, :shipping_method_id, :special_instructions, :use_billing
```
Comment (dkubb) [inline core/lib/spree/permitted_attributes.rb]: This attribute should probably use `%i[...]`.

Should all the lists of permitted attributes be fixed in their own commit?

File: core/spec/models/spree/promotion_handler/cart_spec.rb [hunk 8]
```diff
@@ -51,7 +51,7 @@ module PromotionHandler
       end
 
       context "activates promotions associated with the order" do
-        let(:promo) { create :promotion_with_item_adjustment, adjustment_rate: 5, code: 'promo' }
+        let(:promo) { create :promotion_with_item_adjustment, adjustment_rate: 5 }
```
Comment (dkubb) [inline core/spec/models/spree/promotion_handler/cart_spec.rb]: This should be parenthesized.

File: api/app/controllers/spree/api/promotions_controller.rb (hunk 2 above)
Comment (mbj) [inline api/app/controllers/spree/api/promotions_controller.rb]: @dkubb As the DB can never contain a `NULL` value for `id` we can filter this call. `@promotion = nil` is handled in the action correctly.

I did not spot this originally imperfect code as I was on the high level task. Good catch.

File: core/app/models/spree/promotion.rb (hunk 6 above)
Comment (mbj) [inline core/app/models/spree/promotion.rb]: Highlevel refactoring blindness, good catch.

**Comment by mbj** [comment]
@dkubb ready for the next pass.

File: api/app/controllers/spree/api/promotions_controller.rb [hunk 9]
```diff
@@ -19,7 +19,9 @@ def requires_admin
         end
 
         def load_promotion
-          @promotion = Spree::Promotion.find_by_id(params[:id]) || Spree::Promotion.with_coupon_code(params[:id])
+          id = params[:id]
+
+          @promotion = Spree::Promotion.find(id) if id
```
Comment (dkubb) [inline api/app/controllers/spree/api/promotions_controller.rb]: if `id` is optional, then it might be nicer to write this as:

```ruby
@promotion = Spree::Promotion.find_by(params.slice(:id))
```

Comment (mbj) [inline api/app/controllers/spree/api/promotions_controller.rb]: Would `find_by` short circut on `id: nil` in the arguments?

File: core/lib/spree/permitted_attributes.rb [hunk 10]
```diff
@@ -33,8 +33,8 @@ module PermittedAttributes
       state: [:name, :abbr]
     ].freeze
 
-    @@checkout_attributes = [
-      :coupon_code, :email, :shipping_method_id, :special_instructions, :use_billing
+    @@checkout_attributes = %i[
```
Comment (dkubb) [inline core/lib/spree/permitted_attributes.rb]: @mbj wdyt about refactoring all the permitted attributes to use `%i[...]` in a follow-up PR?

File: api/app/helpers/spree/api/api_helpers.rb [hunk 11]
```diff
@@ -140,7 +140,7 @@ def required_fields_for(model)
 
       @@promotion_attributes = [
         :id, :name, :description, :expires_at, :starts_at, :type, :usage_limit,
-        :match_policy, :code, :advertise, :path
+        :match_policy, :advertise, :path
```
Comment (dkubb) [inline api/app/helpers/spree/api/api_helpers.rb]: What about using `%i[...]` here too?

File: core/lib/spree/permitted_attributes.rb (hunk 10 above)
Comment (mbj) [inline core/lib/spree/permitted_attributes.rb]: I was close to push back on the original flag for consistency. And I'm fine with a followup.

File: core/app/models/spree/promotion.rb [hunk 12]
```diff
@@ -180,9 +174,7 @@ def blacklisted?(promotable)
     end
 
     def normalize_blank_values
-      [:code, :path].each do |column|
-        self[column] = nil if self[column].blank?
-      end
+      self.path = nil if path.blank?
```
Comment (dkubb) [inline core/app/models/spree/promotion.rb]: You can probably refactor this to:

```ruby
self.path = path.presence
```

**Comment by dkubb** [comment]
@mbj I finished a pass over this PR.

File: api/app/helpers/spree/api/api_helpers.rb (hunk 11 above)
Comment (mbj) [inline api/app/helpers/spree/api/api_helpers.rb]: @dkubb What about not using `%i[]` and later do a dedicated pass. Or add a commit that does all of them in this file. But breaking the consistency feels odd.

Comment (dkubb) [inline api/app/helpers/spree/api/api_helpers.rb]: @mbj actually, earlier I was suggesting making a second commit that does all these changes at once.

File: api/app/controllers/spree/api/promotions_controller.rb (hunk 9 above)
Comment (mbj) [inline api/app/controllers/spree/api/promotions_controller.rb]: This does still trigger the redundant `IS NULL` query:

```
[1] pry(main)> Spree::Promotion.find_by(id: nil)
  Spree::Promotion Load (1.6ms)  SELECT  "spree_promotions".* FROM "spree_promotions"  WHERE "spree_promotions"."deleted_at" IS NULL AND "spree_promotions"."id" IS NULL LIMIT 1
=> nil
```

**Comment by mbj** [comment]
Depends on #214.

File: core/app/models/spree/order.rb [hunk 13]
```diff
@@ -544,14 +543,6 @@ def state_changed(name)
       end
     end
 
-    def coupon_code=(code)
-      @coupon_code = code.strip.downcase rescue nil
```
Comment (dkubb) [inline core/app/models/spree/order.rb]: Ugh, this code is horrible, so glad we are removing it.

**Comment by dkubb** [comment]
I am fine with the code in this PR. I will merge once #214 is merged, and this is rebased.

File: core/app/models/spree/promotion.rb [hunk 14]
```diff
@@ -28,22 +28,16 @@ class Promotion < Spree::Base
 
     before_save :normalize_blank_values
```
Comment (dkubb) [inline core/app/models/spree/promotion.rb]: Should this be done in `before_validation` instead?

It's slightly earlier in the process. In general I believe transformations should be pushed as early as possible in the data flow pipeline, so that the "final" (or closest to final) value is available as early as possible within the system.

This should also allow you to change the spec so that instead of `create(:promotion)` you could probably just use `build(:promotion)`, removing the need to persist the record to test normalization.

File: core/spec/models/spree/promotion_spec.rb [hunk 15]
```diff
@@ -522,22 +512,11 @@
   # admin form posts the code and path as empty string
   describe "normalize blank values for code & path" do
     it "will save blank value as nil value instead" do
-      promotion = create(:promotion, code: '', path: '')
-      expect(promotion.code).to be(nil)
+      promotion = create(:promotion, path: '')
```
Comment (dkubb) [inline core/spec/models/spree/promotion_spec.rb]: Assuming you take my advice in https://github.com/MountainRoseHerbs/spree/pull/210/files#r61780963 this can be changed to `build(:promotion, path: '')`.

File: core/app/models/spree/promotion.rb (hunk 14 above)
Comment (mbj) [inline core/app/models/spree/promotion.rb]: I do not want to mix "pure removals" with semantic improvements at this point. Also because more removals to adjacent subystems are in the queue that might kill the entire call side of the proposed fixed seamtntics.

In this case #232 will remove it.

Comment (dkubb) [inline core/app/models/spree/promotion.rb]: Ok. I am fine with this reasoning.

**Comment by dkubb** [comment]
@mbj I am fine to merge this on green.
